### PR TITLE
Add instruction for polymorphism relations in PromptBuilder

### LINKED ISSUE #254: If the prompt only has polymorphic relationships, the prompt does not provide any instruction
**Describe the bug**
In the case that we don't have any method signatures in the prompt, but instead we have multiple polymorphic relations, the prompt just provide sentences like "Class X is a sub-class of class Y". In these cases, LLM usually mocks these classes and ignores the fact that it can use the sub-classes in the test.

**To Reproduce**
1. Generate tests that do not have any objects as input parameters of their methods (e.g., [this class](https://github.com/pderakhshanfar/TestAssignment/blob/main/src/main/java/org/jetbrains/car/Car.java))
2. check the prompt

**Expected behavior**
The `insertPolymorphismRelations` method should also add instructions like `Use the following information about the polymorphic relationship of classes that exist in the project. Don't use mocking for these classes`.

### PR DESCRIPTION
**Pull request by pderakhshanfar**
# Description of changes made
Add instruction for polymorphism relations in PromptBuilder
# Why is merge request needed
Better test generation with less mocking

# Other notes
Closes #254

- [x] I have checked that I am merging into correct branch

### CODE REVIEW

File: core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt [hunk 1]
```diff
@@ -118,6 +118,10 @@ internal class PromptBuilder(private var prompt: String) {
                     fullText += "${currentSubClass.qualifiedName} is $subClassTypeName ${entry.key.qualifiedName}.\n"
                 }
             }
+            // If we have some information added for polymorphism relations, we add an instruction
+            if (fullText.isNotBlank()) {
+                fullText = "Use the following information about polymorphic relationship of classes that exist in the project. Don't use mocking for these classes:\n$fullText" // TODO
```
Comment (Vladislav0Art) [inline core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt:123]: I cannot understand what exactly is instructed in this prompt. Could you explain? **(1)**

If I understood correctly the problem mentioned in the [issue](https://github.com/JetBrains-Research/TestSpark/issues/254), LLM thinks that the entry `class X is a sub-class of Y` means that there is no class `X`, and we want the LLM to create this `X` and make it a sub-class of `Y`. Is it correct? **(2)**

There are several questions:
  1. Is this prompt required to emphasize that the classes `X` are **present in the project** and the LLM **must not create mocks** for classes `Y`? **(3)**
2. `Don't use mocking for these classes` - for what classes: classes `X` or classes `Y`? **(4)**

How about something like this:
_"Consider the following polymorphic relationships of classes present in the project. Use them for instantiation when necessary. Do not mock classes if an instantiation of a sub-class is applicable"_ 
OR:
_"Use the following polymorphic class relationships [present in the project] for instantiation where applicable. Avoid mocking when a subclass instantiation is possible"_

You may use **(i)** to refer to questions.

File: core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt [hunk 2]
```diff
@@ -118,6 +118,10 @@ internal class PromptBuilder(private var prompt: String) {
                     fullText += "${currentSubClass.qualifiedName} is $subClassTypeName ${entry.key.qualifiedName}.\n"
                 }
             }
+            // If we have some information added for polymorphism relations, we add an instruction
+            if (fullText.isNotBlank()) {
```
[the remark this answers is not in the document]
Comment (pderakhshanfar) [reply core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt]: Addressed in my new comment

File: core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt (hunk 1 above)
Comment (pderakhshanfar) [reply core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt]: (1) Your following suggestions for instruction are more reflective, so I have updated them.
(2) and (3) Yes. Without instruction, it just mocks the class and ignores the sub-class. With not mentioning it is available in the project, it often generates both the sub-class and the tests.
(4) for sub class. In general, your suggestions are better. I updated the instructions according to them. Thanks :)

File: core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt [hunk 3]
```diff
@@ -106,8 +106,11 @@ internal class PromptBuilder(private var prompt: String) {
     ) = apply {
         val keyword = "\$${PromptKeyword.POLYMORPHISM.text}"
         if (isPromptValid(PromptKeyword.POLYMORPHISM, prompt)) {
-            var fullText = ""
-
+            // If polymorphismRelations is not empty, we add an instruction to avoid mocking classes if an instantiation of a sub-class is applicable
+            var fullText = when {
+                polymorphismRelations.isNotEmpty() -> "Use the following polymorphic relationships of classes present in the project. Use them for instantiation when necessary. Do not mock classes if an instantiation of a sub-class is applicable"
```
Comment (Vladislav0Art) [inline core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt:111]: I remember some prompt parts were stored in a file with all the string literals (and then accessed as `TestSparkBundle.getString("variable-name-of-string")`. Do we still do it or having a prompt part within the source code is fine?

**Review by Vladislav0Art — APPROVED** [review]

File: core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt (hunk 3 above)
Comment (arksap2002) [reply core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptBuilder.kt:111]: we are doing it everwhere yet, I suggest to create an issue and fix it in a separate PR

**Review by arksap2002 — APPROVED** [review]

**Comment by Vladislav0Art** [comment]
Should we merge it? The PR has hung for a while now.

**Pull request merged**